stats: improve handling of some expected errors

yuzefovich opened the pull request:
In 25.1 we enabled automatic partial stats collection by default. It
works by scanning an index of the table outside of the bounds according
to the last stats collection (i.e. lower than min and higher than max).
If the last stats collection was such that there is no lower bound (which
can happen when we only have a bucket for NULL values as well as "outer"
buckets), we fail the partial stats collection. Previously, this would
result in failing the stats collection job which is logged at the ERROR
level, but this failure is benign and shouldn't result in a scary
looking log.

To prevent unnecessary user concerns this commit adds the logic to
swallow the benign errors which allows us to short-circuit the
particular stats collection without failing the job. We currently treat
all errors with `ObjectNotInPrerequisiteState` PG code as "benign"
(there are 4 more related to partial stats, apart from the scenario
mentioned above). To have some observability into this swalling behavior
we add a log message at INFO level. Additionally, we don't swallow the
error if the stats collection was executed via EXPLAIN or EXPLAIN ANALYZE
(meaning that it was explicitly user-initiated).

The fact that the job will be marked as having "succeeded" can be
confusing since it won't have a corresponding new statistics, but we find
this behavior better overall.

Additionally, this commit extends the error message to also include the
table and index names as well as their IDs.

Fixes: https://github.com/cockroachlabs/support/issues/3360.
Epic: None

Release note (bug fix): In 25.1 version CockroachDB enabled automatic
partial stats collection by default (by flipping
`sql.stats.automatic_partial_collection.enabled` cluster setting to
`true`). Previously, partial stats collection could encounter some
expected scenarios that would be reported with PG error code `55000`
and would show up as failed stats jobs. These errors are benign and will
no longer be reported (the stats job will be marked as "succeeded" but
new stats won't be created).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/149279)
<!-- Reviewable:end -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thank you!

> Reviewed 8 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149279#-:-OU5I68WDdagB5mhjmIO:b-ae66e1)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2 and @mw5h)

___
*[`pkg/sql/stats/bounds/extremes.go` line 80 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149279#-OU5Gk0CAjAdXBbWnzNW:-OU5Gk0CAjAdXBbWnzNX:bf53777) ([raw file](https://github.com/cockroachdb/cockroach/blob/067898bcf7c74df9ab8d8efe1d0fb263645c30d5/pkg/sql/stats/bounds/extremes.go#L80)):*
> ```Go
> 
> // GetUsingExtremesBounds returns a tree.Datum representing the exclusive upper
> // and exclusive lower bounds of the USING EXTREMES span for partial statistics.
> ```

nit: explain that nil may be returned for either value if an exclusive bound cannot be determined from the histogram.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
@mw5h is this similar to what you are doing for #148413?

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149279#-:-OU5ugaZBE8fg9Bp-UGP:b-1hbkk8)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mw5h and @yuzefovich)

___
*[`-- commits` line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149279#-OU5u1DF3csZmj79B7Bq:-OU5tzna2mhZODvgL06O:b-qjeq7m):*
IMO we should add a release note, since this has been observed by customers.
___
*[`pkg/jobs/errors.go` line 169 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149279#-OU5vI_J4cqX1KkAZL56:-OU5vI_J4cqX1KkAZL57:bki87on) ([raw file](https://github.com/cockroachdb/cockroach/blob/067898bcf7c74df9ab8d8efe1d0fb263645c30d5/pkg/jobs/errors.go#L169)):*
> ```Go
> // isBenignStatsError returns whether the given error should be treated as
> // benign (i.e. expected to occur) for table stats related jobs.
> func isBenignStatsError(jobType jobspb.Type, err error) bool {
> ```

It's a bit of a layer violation for the jobs system to have autostats-specific logic.

I'm not sure this is the right approach. @msbutler or @dt, do you have any thoughts about this? We can start a slack conversation if that's easier.


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149279)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @msbutler, @mw5h, and @yuzefovich)

___
*[`pkg/jobs/errors.go` line 169 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/149279#-OU5vI_J4cqX1KkAZL56:-OU5wqJ85zZKS4cjR9zX:b-7z3g8e) ([raw file](https://github.com/cockroachdb/cockroach/blob/067898bcf7c74df9ab8d8efe1d0fb263645c30d5/pkg/jobs/errors.go#L169)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

It's a bit of a layer violation for the jobs system to have autostats-specific logic.

I'm not sure this is the right approach. @msbutler or @dt, do you have any thoughts about this? We can start a slack conversation if that's easier.
</blockquote></details>

thanks for the heads up!

Yeah, this feels like the wrong place for job-type-specific error classification and handling: if the stats jobs believes some class of error is not a failure, it should inspect that error inside its `Resume` method and elect not to return a non-nil error to the job registry rather than asking the job registry to make value judgments about one error vs another. 

If you're concerned that your `Resume()` could return what it considers to be benign errors in many places of different branches, perhaps renaming it to `doResume()` and wrapping that in a call and err check in the public `Resume` could be the answer, but I don't like pushing this logic up into the generic jobs runner code.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
I was thinking more from the perspective that if we don't return an error, then the job is marked as "succeeded", which can be confusing later on since we actually won't find new stats that correspond to that job. In other words, it'd break the assumption (that at least I have) that "successful auto stats job = new table stats". Perhaps this will be quite rare and adding an info log for this would be sufficient for our needs? @michae2 I think you mentioned the same concern, WDYT?

dt reviewed (commented):
Well you could keep the successful job => new stats implication if you avoided creating jobs in the first place that you know will fail to create new stats due to "benign" errors. But either the job "succeeded"  -- what that means is up to it, and you can say it means it made new stats or you can say it either made new stats or actually decided not to because of a "benign error" -- or it failed. If it failed. We should one of increment the succeeded or failed metric here (or cancelled), because it is one of those three things -- I don't really want to say "it sort of failed, but it is a special failure that means the jobs system metrics don't add up and we log differently".

If you _really_ want to keep the structure where jobs are, shall we say, speculatively created and Resume()'ed with the expectation that they might not succeed but have their somewhat expected lack of success be classed as less concerning than normal failures, I could get behind allowing a job to self-request cancelation? We already allow jobs to self-request pausing, by returning a PauseRequested error. If you wanted to introduce a similar CancellationRequested error, a job could elect to return that if it determines that it would like to cancel itself rather than succeed (return no error) or fail (return an error)?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149279#-:-OU6JzgED9MRgScTUKHS:b-gmn56d)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @msbutler, @mw5h, and @yuzefovich)



<!-- Sent from Reviewable.io -->

michae2 commented:
> In other words, it'd break the assumption (that at least I have) that "successful auto stats job = new table stats". Perhaps this will be quite rare and adding an info log for this would be sufficient for our needs? @michae2 I think you mentioned the same concern, WDYT?

Yes, this is the other side of the coin. It's potentially confusing, but on reflection I think it's less confusing than "job error != serious error" since people have built monitoring around errors / job failures and want to understand each one.

> If you wanted to introduce a similar CancellationRequested error, a job could elect to return that if it determines that it would like to cancel itself rather than succeed (return no error) or fail (return an error)?

Cool idea! This is slick, but also I'm not sure how backportable it would be.

IMO we should swallow the error inside the job, potentially log it as INFO or WARN, and not report it to the jobs system.

yuzefovich committed:
stats: improve handling of some expected errors

In 25.1 we enabled automatic partial stats collection by default. It
works by scanning an index of the table outside of the bounds according
to the last stats collection (i.e. lower than min and higher than max).
If the last stats collection was such that there is no lower bound (which
can happen when we only have a bucket for NULL values as well as "outer"
buckets), we fail the partial stats collection. Previously, this would
result in failing the stats collection job which is logged at the ERROR
level, but this failure is benign and shouldn't result in a scary
looking log.

To prevent unnecessary user concerns this commit adds the logic to
swallow the benign errors which allows us to short-circuit the
particular stats collection without failing the job. We currently treat
all errors with `ObjectNotInPrerequisiteState` PG code as "benign"
(there are 4 more related to partial stats, apart from the scenario
mentioned above). To have some observability into this swalling behavior
we add a log message at INFO level. Additionally, we don't swallow the
error if the stats collection was executed via EXPLAIN or EXPLAIN ANALYZE
(meaning that it was explicitly user-initiated).

The fact that the job will be marked as having "succeeded" can be
confusing since it won't have a corresponding new statistics, but we find
this behavior better overall.

Additionally, this commit extends the error message to also include the
table and index names as well as their IDs.

Release note (bug fix): In 25.1 version CockroachDB enabled automatic
partial stats collection by default (by flipping
`sql.stats.automatic_partial_collection.enabled` cluster setting to
`true`). Previously, partial stats collection could encounter some
expected scenarios that would be reported with PG error code `55000`
and would show up as failed stats jobs. These errors are benign and will
no longer be reported (the stats job will be marked as "succeeded" but
new stats won't be created).

yuzefovich reviewed (commented):
Thanks for the feedback! I refactored the code to swallow the error in the code path that is used by the auto stats job, and if we swallow the error, we'll now log it at INFO level. Additionally, we have a separate EXPLAIN / EXPLAIN ANALYZE path where we keep the current behavior of reporting an error to the client.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149279#-:-OU7og5hBDxd8u1d3vRp:b-9xvmxa)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @mw5h)



<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
Great discussion above! I like where this ended up and I think it is more intuitive than before. When the benign error occurs the job was successful, i.e., it completed and should not be retried. The fact that there was no outcome from the stats job is specific to the type of job—stats—not an indication of a general job failure.

> Reviewed 11 of 11 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149279#-:-OUAOGOG0WQzzLQiijXd:ba7xpsx)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mw5h)



<!-- Sent from Reviewable.io -->

rytaft commented:
>  I'd lean towards backporting it to 25.1 and 25.2 with default "fix=on"

With the new policy, default will need to be off for 25.1

michae2 reviewed (approved):
LGTM. Thank you for doing this!

mw5h reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 8 files at r1, 11 of 11 files at r2, 2 of 2 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/149279#-:-OUHeynj--GpIvG0egx0:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
Now let's agree on how we should be backporting this. I still think that a version gate is not warranted here.

>> I'd lean towards backporting it to 25.1 and 25.2 with default "fix=on"

> With the new policy, default will need to be off for 25.1

25.1 is an Innovation release that will never be LTS, so the bar to backporting there seems a bit lower than for Regular releases. From the updated policy I think it would fall under "inability to detect ... production issues" under "fixed for serious issues". Well, I guess this ensures the _need_ for the backport rather than the _shape_ of the backport. If we backport this to 25.1 with "off" by default, then we're bound to get more inquiries from customers since over the years we've taught them to pay attention to failed stats collections. I imagine that one of our larger customers, if they don't skip 25.1 release, is bound to complain similar to https://github.com/cockroachlabs/support/issues/3359 (although we will have the knob by that time). Still, the risk of enabling it by default seems minuscule in comparison to possible alerts customers will get (and then us getting support tickets), so I think we should backport as is, without the version gate, the patch always enabled. Thoughts?

yuzefovich commented:
blathers auth-check
